### PR TITLE
Add overset and underset support for mathtext

### PR DESCRIPTION
**Pull request by aitikgupta**
## PR Summary
This PR addresses #18241, adds `\overset` and its variant `\underset` LaTeX symbol, which looks like this:
- Overset:
mathtext_cm_83
- Underset:
mathtext_cm_84

Also see: http://tex.wikidot.com/snippets:overset-and-underset

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by aitikgupta** [comment]
I followed the developer guide for image comparisons, [ref](https://matplotlib.org/devel/testing.html#writing-an-image-comparison-test), and I pass the tests locally.
Although, going through the `result_images`, I find there weren't any `svg` extensions generated for the test I appended - It only generated `png` and `pdf` extensions.
For all the previous `baseline_images`, there are `svg`s along with `png`s and `pdf`s.

Is there something I missed, since previous PRs related to `mathtext` do have `svg`s in them? For example, https://github.com/matplotlib/matplotlib/pull/8151.

**Comment by aitikgupta** [comment]
Hey @QuLogic, could you take a look at this too? Thanks :D

File: lib/matplotlib/_mathtext.py [hunk 1]
```diff
@@ -2858,6 +2876,42 @@ def binom(self, s, loc, toks):
...
+                           cvalue                       # value
+                           ])
+
+            # Shift so the symbol sits in the same horizontal position
+            metrics = state.font_output.get_metrics(
+                state.font, mpl.rcParams['mathtext.default'],
+                r'\rightarrow', state.fontsize, state.dpi)
+
+            shift = ((csymbol.height) / 2. +
+                     (metrics.ymax + metrics.ymin) / 2. + thickness * 2.)
```
Comment (QuLogic) [inline lib/matplotlib/_mathtext.py:2903]:
```suggestion
            shift = (csymbol.height / 2 +
                     (metrics.ymax + metrics.ymin) / 2 + thickness * 2)
```

Comment (QuLogic) [reply lib/matplotlib/_mathtext.py:2903]: I'm not sure what's wrong here, but it appears that `underset` is a little bit lower than `overset`:
```python
import matplotlib.pyplot as plt
plt.text(0.5, 0.5, r'Test of some $math \overset{foo}{\rightarrow} \underset{bar}{\rightarrow}$')
plt.savefig('foo.png', dpi=600)
```
image

Comment (aitikgupta) [reply lib/matplotlib/_mathtext.py:2903]: You're right, there were some logistic issues with the old approach:
Screenshot from 2020-12-09 16-46-53
I made some changes, this is how it looks now:
Screenshot from 2020-12-09 16-40-52
Alternatively, it'd be helpful to look at some characters which are not italicised or have all characters of same height:
Screenshot from 2020-12-09 16-43-24
Screenshot from 2020-12-09 16-50-30

File: lib/matplotlib/_mathtext.py [hunk 2]
```diff
@@ -2858,6 +2876,40 @@ def binom(self, s, loc, toks):
...
+        csymbol = HCentered([symbol])
+        width = max(cvalue.width, csymbol.width)
+        cvalue.hpack(width, 'exactly')
+        csymbol.hpack(width, 'exactly')
+
+        metrics = state.font_output.get_metrics(
+            state.font, mpl.rcParams['mathtext.default'],
+            r'\leftrightarrow', state.fontsize, state.dpi)
+
+        # Shift so the symbol sits in the same horizontal position
```
Comment (QuLogic) [inline lib/matplotlib/_mathtext.py:2895]: vertical position, isn't it?

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/_mathtext.py (hunk 2 above)
Comment (aitikgupta) [reply lib/matplotlib/_mathtext.py:2895]: Ah, that's right

**Comment by aitikgupta** [comment]
Removed the extra commits

File: lib/matplotlib/tests/test_mathtext.py [hunk 3]
```diff
@@ -109,6 +109,8 @@
     r'$,$ $.$ $1{,}234{, }567{ , }890$ and $1,234,567,890$',  # github issue 5799
     r'$\left(X\right)_{a}^{b}$',  # github issue 7615
     r'$\dfrac{\$100.00}{y}$',  # github issue #1888
+    r'$x \overset{f}{\rightarrow} y$',  # github issue #18241
+    r'$x \underset{f}{\rightarrow} y$',  # github issue #18241
```
Comment (anntzer) [inline lib/matplotlib/tests/test_mathtext.py:113]: Please consider merging these two tests together into a single test case, as this should result in a nearly twofold save in file size for the vector formats (which won't have to encode the glyphs twice).
The commits should be squashed after merging the test cases to remove the separate files from the git history.

Comment (aitikgupta) [reply lib/matplotlib/tests/test_mathtext.py:113]: That makes sense, so after merging it would look like:
`r'$x \overset{f}{\rightarrow} \underset{f}{\rightarrow} y$'`

Comment (anntzer) [reply lib/matplotlib/tests/test_mathtext.py]: thanks

**Comment by timhoffm** [comment]
Doc build failure seems due to #19163. @aitikgupta could you please rebase on master to pick up that change, so that docs build cleanly?

@anntzer I'd like to have your review as our "master of fonts" :smile:. Feel free to turn the request down if you don't have time.

File: lib/matplotlib/_mathtext.py [hunk 4]
```diff
@@ -2168,6 +2172,18 @@ def __init__(self):
             - (p.required_group | Error("Expected \\overline{value}"))
         )
 
+        p.overset <<= Group(
+            Suppress(Literal(r"\overset"))
+            - ((p.required_group + p.symbol_group)
```
Comment (anntzer) [inline lib/matplotlib/_mathtext.py:2177]: Should this not just use simple_group instead?  (in which case symbol_group could go away)  Looks like TeX is perfectly happy with `\overset{ab}{}` or `\overset{ab}{cd}` (you may need to add tests that both "overset over nothing" and "overset over multiple characters" work).

File: lib/matplotlib/_mathtext.py [hunk 5]
```diff
@@ -2901,6 +2953,30 @@ def overline(self, s, loc, toks):
...
 
+    def overset(self, s, loc, toks):
+        assert len(toks) == 1
+        assert len(toks[0]) == 2
+        # Only one symbol
+        assert len(toks[0][1]) == 1
+        state = self.get_state()
+
+        value, symbol = toks[0]
+        symbol = symbol[0]
```
Comment (anntzer) [inline lib/matplotlib/_mathtext.py:2964]: you can just write e.g. `(value, (symbol,)), = toks` and get rid of the asserts (and likewise for `underset`.

File: lib/matplotlib/_mathtext.py [hunk 6]
```diff
@@ -2841,6 +2859,40 @@ def binom(self, s, loc, toks):
...
+        cvalue.hpack(width, 'exactly')
+        csymbol.hpack(width, 'exactly')
+
+        metrics = state.font_output.get_metrics(
+            state.font, mpl.rcParams['mathtext.default'],
+            r'\leftrightarrow', state.fontsize, state.dpi)
+
+        # Shift so the symbol sits in the same vertical position
+        shift_amount = csymbol.height/2 - (metrics.ymax + metrics.ymin)/2
+        if shift:
```
Comment (anntzer) [inline lib/matplotlib/_mathtext.py:2880]: given that this is tightly coupled to over/under (as the vlist order is also set here), I'd just rename `shift` to e.g. `overunder` which is either "over" or "under" for readability.

File: lib/matplotlib/_mathtext.py [hunk 7]
```diff
@@ -2841,6 +2859,40 @@ def binom(self, s, loc, toks):
         return self._genfrac('(', ')', 0.0, self._MathStyle.TEXTSTYLE,
                              num, den)
 
+    def _genset(self, state, value, symbol, shift):
```
Comment (anntzer) [inline lib/matplotlib/_mathtext.py:2862]: It's not super obvious which of `value` and `symbol` is the main part and which one is the "annotation", I guess we could rename them to e.g. `body` and `annotation` (throughout).

**Comment by anntzer** [comment]
I didn't really follow @QuLogic's discussion about alignments, but right now looking at
```python
rcParams["text.latex.preamble"] = r"\usepackage{amsmath}"
s = r"$\overset{a}{a}\overset{p}{a}\overset{a}{b}\overset{p}{b}\underset{a}{a}\underset{b}{a}\underset{a}{p}\underset{b}{p}$"
figtext(.5, .6, s, usetex=True)
figtext(.5, .4, s)
```
we get
test
It looks like TeX's approach is to make sure the baselines of the main symbols stay always aligned (which seems more useful than trying to align the decorators?); likely we should do the same?

File: lib/matplotlib/_mathtext.py (hunk 4 above)
Comment (aitikgupta) [reply lib/matplotlib/_mathtext.py]: TeX is also happy with `\overset{}{}`, this made me _remove_ the tests (and also convert the `required_group` into a `simple_group`), since we only test failures:
https://github.com/matplotlib/matplotlib/blob/739898e708ec1dc8228b421e3bfba74a631706dc/lib/matplotlib/tests/test_mathtext.py#L236-L238

**Comment by aitikgupta** [comment]
Looks like what @QuLogic said about `underset` being a bit lower than `overset` is in TeX backend too:
### `usetex=True`


### `usetex=False`


#### With the latest commit, alignment looks like this:

**Comment by anntzer** [comment]
I would think it would be better if the p's baselines could also be aligned?

**Comment by aitikgupta** [comment]
> I would think it would be better if the p's baselines could also be aligned?

Something like this?

**Comment by anntzer** [comment]
Yes.

**Comment by jklymak** [comment]
But thT is still not the same?

**Comment by aitikgupta** [comment]
> But thT is still not the same?

I'm not sure I understand, do you mean the baselines do not match? Maybe this will help:

**Comment by jklymak** [comment]
The overset and underset baselines are not aligned

**Comment by timhoffm** [comment]
They are not aligned in TeX either, but the spread here is a bit larger than in TeX. Personally, I don't find that too bad.

**Comment by aitikgupta** [comment]
The spread is a bit larger even for a normal text:


We could reduce the spread in the final `hlist` theoretically, it will look something like this (have not pushed):


Just for reference, here is the current version (which is currently in the PR):


Should I update the PR to remove the gap?@jklymak @timhoffm

**Comment by timhoffm** [comment]
@aitikgupta I meant the spread of the baselines of the sub/supersets between a and p.
But you are right that the character spread is a bit larger too.

Personally I'm ok with both variants in both cases, but I'm not the most critical one here.

**Comment by anntzer** [comment]
I agree the horizontal gap should, if possible (I guess that means "if the annotation is not too wide") stay the same as without annotation:
```python
rcParams["text.latex.preamble"] = r"\usepackage{amsmath}"
figtext(.5, .50, r"$\underset{a}{b}\underset{a}{b}$"); figtext(.5, .45, r"$bb$")
figtext(.5, .40, r"$\underset{a}{b}\underset{a}{b}$", usetex=True); figtext(.5, .35, r"$bb$", usetex=True)
```
currently gives
test
which is too wide on our side.

Also, the spacing between the decoration and the body seems to small.  A quick search suggests that TeX's overset/underset is defined in terms of `\limits` (https://tex.stackexchange.com/questions/42456/creating-non-math-mode-substitutes-for-overset-and-underset-not-dependent-on-t/42477), so I guess(?) the spacings should be the same as for e.g. `\lim`.  But for `\lim` we use a spacing of 3rules (see `is_overunder` block in `subsuper`), so is there a reason you user 1.5rules for overset/underset?  (Again, I may have missed an earlier discussion.)  (Looking at the TeXbook the actual spacings for `\limits` are defined in terms of internal fontdimens, so we're probably not going to match these exactly.)

**Comment by aitikgupta** [comment]
Okay this makes sense, I will update the PR - will fix the horizontal gap.
About the spacing between body and annotation -  it could be increased (value of `1.5` is hardcoded here; similar to `-3ex` in one of the answers in stackexchange link @anntzer shared):
```latex
\newcommand\textoverset[2]{
  ....
      \noalign{\kern-.3ex}
  ...
}
```
https://github.com/matplotlib/matplotlib/blob/8e533a8214ab0ce72a8d48e07732a42be67f5f95/lib/matplotlib/_mathtext.py#L2874

We can play around with this value; can't have an exact match since:
> Looking at the TeXbook the actual spacings for \limits are defined in terms of internal fontdimens (big_op_spacing*), so we're probably not going to match these exactly.

**Comment by anntzer** [comment]
Some links for reference about TeX's placement algorithms (the figures are nice): https://www.tug.org/TUGboat/tb27-1/tb86jackowski.pdf http://www.tug.org/~vieth/papers/bachotex2008/math-font-paper.pdf https://www.ntg.nl/maps/38/03.pdf
I'm absolutely not asking that this PR implements any of the algorithms described in them, but they can be useful as reference targets.

**Comment by aitikgupta** [comment]
These links were amazing reads!
The latest push is a bit inspired from these resources.
@anntzer can you review `_genset` block again? I think it'll make a bit more sense now since `Vbox` is out of the picture.

File: lib/matplotlib/_mathtext.py [hunk 8]
```diff
@@ -2841,6 +2857,38 @@ def binom(self, s, loc, toks):
...
+        cbody = HCentered([body])
+        cannotation = HCentered([annotation])
+        width = max(cbody.width, cannotation.width)
+        cbody.hpack(width, 'exactly')
+        cannotation.hpack(width, 'exactly')
+
+        kern_vgap = thickness * 3
+        if overunder == "under":
+            vlist = Vlist([cannotation,                # annotation
+                           Kern(kern_vgap),            # space
```
Comment (anntzer) [inline lib/matplotlib/_mathtext.py:2875]: Can you (for my personal education) explain why this needs to be a Kern and not a Vbox?  Does that mean the other places that use Vboxes (e.g. `_genfrac`, `c_over_c`) should also be changed?

File: lib/matplotlib/_mathtext.py [hunk 9]
```diff
@@ -2841,6 +2857,38 @@ def binom(self, s, loc, toks):
...
+            shift_amount = cannotation.depth + cbody.height + kern_vgap
+
+            vlist.shift_amount = shift_amount
+        else:
+            vlist = Vlist([cbody,                      # body
+                           Kern(kern_vgap),            # space
+                           cannotation                 # annotation
+                           ])
+
+        hlist = [Hlist([vlist])]
```
Comment (anntzer) [inline lib/matplotlib/_mathtext.py:2888]: Do you need to wrap the return value in a list containing an Hlist?  Can you not just return the Vlist directly?  That's what e.g. c_over_c does.
(Again, I don't know the correct answer...)

File: lib/matplotlib/_mathtext.py (hunk 8 above)
Comment (aitikgupta) [reply lib/matplotlib/_mathtext.py:2875]: We can understand `Kern` as a superset of `Vbox` and `Hbox` - `Vbox` is only used for vertical spaces, `Hbox` is only used for horizontal spaces.
`Kern` can be used in both - it is a more LaTeX-like keyword, so anybody who has an understanding of general LaTeX should feel at home looking at this.
On the other hand, `Vbox` should make more sense in vertical lists - since it is easy to understand what it would do.

Other than this, there is _no difference_ if we change `Kern` to a `Vbox` here, or other places (eg. `_genfrac`, `c_over_c`).

Comment (aitikgupta) [reply lib/matplotlib/_mathtext.py:2875]: Since we wouldn't want a _negative_ `kern_vgap`, I think it's better to stick to `VBox` here?

File: lib/matplotlib/_mathtext.py (hunk 9 above)
Comment (aitikgupta) [reply lib/matplotlib/_mathtext.py:2888]: For simple purposes of readability, and if we ever wanted there to be a horizontal gap between the current symbol and the next one - we'd use:
```python
hlist = [Hlist([vlist, Hbox(0, horizontal_gap)])]
```

File: lib/matplotlib/_mathtext.py (hunk 8 above)
Comment (anntzer) [reply lib/matplotlib/_mathtext.py:2875]: Let's stick to Vbox for consistency with the other constructs (genfrac, c_over_c, etc.)?

File: lib/matplotlib/_mathtext.py (hunk 9 above)
Comment (anntzer) [reply lib/matplotlib/_mathtext.py:2888]: I'd rather not wrap things into more things unless needed; if you want I guess you can leave the note about horizontal gaps as a comment?

Comment (aitikgupta) [reply lib/matplotlib/_mathtext.py]: I made the change, although, I found other places where this is done too:
https://github.com/matplotlib/matplotlib/blob/0520035717925131666ad5bb2ef7e21ec61b40fd/lib/matplotlib/_mathtext.py#L2684-L2687

Comment (anntzer) [reply lib/matplotlib/_mathtext.py]: I guess these can be cleaned up as a separate PR?

**Comment by anntzer** [comment]
Perhaps also modify the test images to include the cases you improved in the last round of review? (e.g. add `\overset{x}{f}\overset{x}{f}\overset{xx}{ff}` etc. to check both horizontal and vertical spacings.)
(the vector files should hardly increase in size as what takes the most space should actually be glyph embedding, but as long as you stick to the same characters this doesn't expand...)

Then squash, and other than that this looks great :)

**Review by anntzer — APPROVED** [review]
Great work, and thanks for your patience handling the reviews :)  I'll give it a day in case anyone else wants to comment but other than that, anyone can merge.
(Also, you can consider adding a whatsnew entry.)

**Comment by anntzer** [comment]
@timhoffm Given your comment at https://github.com/matplotlib/matplotlib/pull/19186#issue-546268424, do you want to comment wrt. baseline images?  (Everything else looks great to me.)

**Comment by timhoffm** [comment]
TL;DR For the sake of simplicity, you can just merge with the new baseline images.

Here it's "only" a little over 40kB of baseline images and the positioning here is much more complex. I think this should be tested and adding the baseline images is okish. (The alternative would be splitting the tests into a separte PR and keep that open until we have less data heavy ways of doing the test - either with separated baseline images or by only using a subset; probably we don't need all combinations of fonts and output formats?).

Btw. what's the status with separating baseline images from the code repo?

**Comment by anntzer** [comment]
I agree positioning matters here much more than for the sqrt PR (for which I agree we could do without the test images...).

> Btw. what's the status with separating baseline images from the code repo?

Some day I'll have time to get back to it.  But that day is not particularly close :(

**Comment by anntzer** [comment]
@aitikgupta Actually let's see whether we can resolve https://github.com/matplotlib/matplotlib/pull/19186#issuecomment-752266330 first to save a lot of disk space on the baseline images problem.  However I don't want you to believe that I'm needlessly temporizing on this so regardless of whether the proposal at https://github.com/matplotlib/matplotlib/pull/19186#issuecomment-752266330 gets implemented I will make sure this is merged before 3.4 is released (hence the self-assignment).

**Review by anntzer — CHANGES_REQUESTED** [review]
Actually https://github.com/matplotlib/matplotlib/pull/19201 turned out to quite simple to implement, so let's decide on that PR first.

**Review by QuLogic — APPROVED** [review]
For whenever we figure out test images.

**Review by anntzer — CHANGES_REQUESTED** [review]
@aitikgupta this needs to be rebased to use the new (cheaper) baseline images mechanism.  (I can do the rebase if you prefer, just let me know.)

**Comment by aitikgupta** [comment]
> (I can do the rebase if you prefer, just let me know.)

No that's fine, I'll do it

**Comment by aitikgupta** [comment]
I think just rebasing wouldn't work, I'd need to generate the baseline images too?

**Comment by anntzer** [comment]
Yes, sorry I forgot to mention that.

**Comment by aitikgupta** [comment]
Is it fine the way it is now?
Or should I consider breaking down the current test into 2 (one for overset, one for underset), since the tests only compare a single png now

**Review by anntzer — APPROVED** [review]
Let's still stick to a single test, it's also useful to be able to compare the alignment behaviors of over and underset easily.

**Comment by anntzer** [comment]
Thanks for the PR :)

**Comment by aitikgupta** [comment]
Uh, while rebasing the last time to add the new baseline images, one of my commits vanished: the whatsnew entry.
It was this commit I suppose: https://github.com/matplotlib/matplotlib/commit/0a030d32fd7d26b1fd48c07028b2f10bf70b238d

Now this is merged, but without the whatsnew entry.
Should we do something about it?
@anntzer @QuLogic

**Comment by anntzer** [comment]
Reopen a separate PR.
